vagrant-pxe-harvester/Vagrantfile

```
@@ -73,9 +73,17 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
         boot_network = {'network' => 'harvester'}
         libvirt.boot 'hd'
         libvirt.boot boot_network
-        # NOTE: default to UEFI boot. Comment this out for legacy BIOS.
-        libvirt.loader = '/usr/share/qemu/ovmf-x86_64.bin'
-        libvirt.nic_model_type = 'virtio'
+
+        # NOTE: UEFI boot if ovmf there, or adopt legacy BIOS.
+        ovmf_ubuntu = '/usr/share/qemu/OVMF.fd'
+        ovmf_opensuse = '/usr/share/qemu/ovmf-x86_64.bin'
+        if File.exist?(ovmf_ubuntu)
+            libvirt.loader = ovmf_ubuntu
+            libvirt.nic_model_type = 'e1000'
+        elsif File.exist?(ovmf_opensuse)
+            libvirt.loader = ovmf_opensuse
+            libvirt.nic_model_type = 'e1000'
```

Shouldn't this be `virtio` on openSUSE?
```suggestion
            libvirt.nic_model_type = 'virtio'
```

---

Good question. I've always used `e1000` even on openSUSE. I just found https://github.com/harvester/ipxe-examples/commit/c427af7 from Kiefer back in 2021 which says virtio doesn't play nice with bonding - I'm not sure if that's changed since...?

---

I'm only asking because it has been changed with commit https://github.com/harvester/ipxe-examples/commit/e34033880b72756ed75f878adc6afa627bf1daad#diff-b6d2e72b1acfe3fb9bdbb6bb4f0b69e31ab586d6d9fc44ae4c8aa06dedee0770R78 and wondering why it has been changed back to `e1000`.
If this was with reason, it might make sense to move `libvirt.nic_model_type = 'e1000'` out of the if else block.

---

Yes, virtio has less overhead for the hypervisor and generally results in better performance in most cases. But there are things that an emulated device does better than a paravirtualized one. Namely guest OS driver support. Windows for example doesn't have drivers for virtio devices out of the box.
But some features of physical nics are also not available on virtio devices or depend on the hypervisor setup (e.g. VLAN).
Additionally, virtio devices aren't a panacea for performance, as it is possible to have bad performance with them if they are set up wrong.

---

> Yes, virtio has less overhead for the hypervisor and generally results in better performance in most cases. But there are things that an emulated device does better than a paravirtualized one. Namely guest OS driver support. Windows for example doesn't have drivers for virtio devices out of the box. But some features of physical nics are also not available on virtio devices or depend on the hypervisor setup (e.g. VLAN). Additionally, virtio devices aren't a panacea for performance, as it is possible to have bad performance with them if they are set up wrong.

Like @m-ildefons mentioned,
so I decided to focus on the different OS path handling and keep the `nic_model_type` the original value. We can open another ticket for it if need.